design/ErrorHandling.md

```diff
@@ -164,147 +163,265 @@ WebGPU access has been disabled for the page.
+It resolves to null if no errors were captured, and otherwise resolves to the first error that occurred in the scope -
+either a `GPUOutOfMemoryError` or a `GPUValidationError` object containing information about the validation failure.
+
+An error scope captures an error if its filter matches the type of the error scope:
+`pushErrorScope('out-of-memory')` captures `GPUOutOfMemoryError`s;
+`pushErrorScope('validation')` captures `GPUValidationError`s.
+`pushErrorScope('none')` never captures errors, but can be used to detect operation completion.
+The filter mechanism prevents developers from accidentally silencing validation errors when trying to do fallible allocation or wait for completion.
+
+If an error scope captures an error, the error is not passed down to the enclosing error scope.
+Each error scope stores only the **first error** it captures, and returns that error when the scope is popped.
+Any further errors it captures are **silently ignored**.
+
+If an error is not captured by an error scope, it is passed out to the enclosing error scope.
+
+If there are no error scopes on the stack, `popErrorScope()` rejects.
+
+If the device is lost, `popErrorScope()` always rejects.
+
+\* Error scope state is **per-device, per-execution-context**.
+That is, when a `GPUDevice` is posted to a Worker for the first time, the new `GPUDevice` copy's error scope stack is empty.
+(If a `GPUDevice` is copied *back* to an execution context it already existed on, it shares its error scope state with all other copies on that execution context.)
+
+```webidl
+enum GPUErrorFilter {
+    "none",
     "out-of-memory",
-    // The object is invalid for another, unrecoverable reason.
-    "invalid",
+    "validation"
 };
+
+interface GPUOutOfMemoryError {};
+
+interface GPUValidationError {
+    readonly attribute DOMString message;
+};
+
+typedef (GPUOutOfMemoryError or GPUValidationError) GPUError;
+
+partial interface GPUDevice {
+    void pushErrorScope(GPUErrorFilter filter);
+    Promise<GPUError?> popErrorScope();
+};
+```
+
+### *Fallible Allocation*
+
+An `out-of-memory` error scope can be used to detect allocation failure.
+
+#### Example: tryCreateBuffer
+
+```js
+async function tryCreateBuffer(device, desc) {
+  device.pushErrorScope('out-of-memory');
+  const buffer = device.createBuffer(desc);
+  if (await device.popErrorScope() !== null) {
+    return null;
+  }
+  return buffer;
+}
+```
+
+### *Waiting for Completion*
+
+Since error scope results only return once the enclosed operations are complete, a `none` error scope can be used to detect completion of off-queue operations.
+(On-queue operation completion can be detected with `GPUFence`.)
+
+#### Example: createReadyRenderPipeline
+
+```js
+async function createReadyRenderPipeline(device, desc) {
+  device.pushErrorScope('none');
+  const pipeline = device.createRenderPipeline(desc);
+  await device.popErrorScope(); // always resolves to null
+  return pipeline;
+}
```

Comment:
The explainer should discourage people from using functions like `createReadyRenderPipeline` in the middle of a bunch of drawing.  The ramifications of async promises + implicit present can be perilous.

Consider the following code:

```js
async void drawFrame() {
    // Draw a tree to the swap chain texture
    let renderPipeline = await createReadyRenderPipeline(...);

    ////// Everything past this point runs outside of the RAF \\\\\\

    // Draw a dog to the swapchain texture
    window.requestAnimationFrame(drawFrame);
}
window.requestAnimationFrame(drawFrame);
```

In the code above, since there is an `await` inside of `createReadyRenderPipeline`, all of the code after the `await` will run in the promise callback of `popErrorScope`.  Since WebGPU has implicit present, your first frame will contain just the tree and your second frame will contain just the dog instead of both the dog and the tree like you expected.  

If you saved away the result of `GPUSwapChain.getCurrentTexture` at the top of the function, you will never see the tree because it draws to the swap chain's previous texture. 

If you wrote the above code in WebXR's RAF, the frame data given to you is no longer valid at draw-dog time so any methods you call on the frame data object will throw exceptions.  You code won't even live to see frame 2.

At a minimum, `createReadyRenderPipeline` should be renamed to `requestReadyRenderPipeline` and any sample code should .`then` the returned promise with code that stores the returned pipeline for use in a subsequent frame.  Only then, will you get expected rendering results in RAF.